add phpstan

garak opened the pull request:
This is a first try to add phpstan.
~~Level is 6 for now, maybe we could increase it eventually.
I'm not really happy with baseline, so I'm open to suggestion about it~~

stof commented on src/Knp/Menu/Util/MenuManipulator.php:
```diff
@@ -276,6 +279,9 @@ private function buildBreadcrumbsArray(ItemInterface $item): array
     }
 
+    /**
+     * @return array<string, string|ItemInterface>
```
`@phpstan-return array{label: string, url: string|null, item: ItemInterface|null}`

stof commented on src/Knp/Menu/Util/MenuManipulator.php:
```diff
@@ -265,6 +265,9 @@ public function getBreadcrumbsArray(ItemInterface $item, $subItem = null): array
     }
 
+    /**
+     * @return array<int, array<string, string|ItemInterface>>
```
`@phpstan-return list<array{label: string, url: string|null, item: ItemInterface|null}>`

stof commented on tests/Knp/Menu/Tests/Loader/ArrayLoaderTest.php:
```diff
@@ -95,6 +95,9 @@ public function testSupports($data, bool $expected): void
     }
 
+    /**
+     * @return array<int, array<int, mixed>>
```
I would not bother doing precise phpdoc for tests IMO

garak replied:
But if we don't do it, phpstan would complain (unless putting everything in baseline, of course)

stof replied:
I'm wondering about the cost/benefit ratio of applying static analysis on the tests folder...

garak replied:
well, the cost is already spent... so I guess that it would cost us more to remove it

stof replied:
it is spent at level 6. but your answer tells me that most of the cost at level max is also in tests. I'd rather analyse only src at max level than lowering the level due to tests.
And this applies also to future changes in tests

garak replied:
I see. Considering this, I have to agree with you.

stof commented:
@garak what would happen if you use phpstan at the max level insteaf of level 6 ?

garak commented:
> @garak what would happen if you use phpstan at the max level insteaf of level 6 ?

for now, we get 292 errors at max level

stof commented:
@garak how many of these in src vs tests ?

garak commented on src/Knp/Menu/Util/MenuManipulator.php:
```diff
@@ -103,9 +103,9 @@ public function slice(ItemInterface $item, $offset, $length = null): ItemInterfa
      * Split menu into two distinct menus.
      *
-     * @param mixed $length name of child, child object, or numeric length
+     * @param string|int|object $length name of child, child object, or numeric length
```
Deprecate `split` and duplicate it in `slice`?
I'll take it for a different PR

stof replied:
no. I'm talking about the replacement of `mixed` with actual supported types

garak commented:
> @garak how many of these in src vs tests ?

With `src` only: 17 errors at level 7, 29 errors at max

stof commented:
@garak can you configure it on the max level now, and generate the remaining baseline (or configure the CI to run it), so that I can also see the list of failures at max level ? It might make sense to go there

stof commented on phpstan-baseline.neon:
```diff
@@ -0,0 +1,127 @@
+			path: src/Knp/Menu/Renderer/ListRenderer.php
+
+		-
+			message: "#^Method Knp\\\\Menu\\\\Renderer\\\\ListRenderer\\:\\:renderLabel\\(\\) should return string but returns string\\|null\\.$#"
```
I think this is caused by a bad phpdoc on `ItemInterface::getLabel()`. The phpdoc says it defaults to the item name, so I don't think it can actually be null.

garak replied:
Do you think we can change interface to return `string` instead of `?string`?

stof replied:
I think using a nullable typehint was a mistake when types were introduced. the phpdoc was saying `string` before that: https://github.com/KnpLabs/KnpMenu/commit/2dd2e7687d23b174689c357cd4a749ff825e4795#diff-8f6d2b955ef3caa9f7aa107c36571bad58fc424afbbc1abdbb4e5a4588d63634R56

garak replied:
I agree. Do you think we can change that return type in a minor version, adding a note in changelog?

stof replied:
given that MenuItem never returns `null` there, I think we can. I don't expect people to have custom implementations of the ItemInterface actually.

garak commented:
By the way, Scrutinizer is reporting a lot of false positives because settings need to be [updated to new version](https://scrutinizer-ci.com/docs/tools/php/php-analyzer/guides/migrate_to_new_php_analysis).
I can't do it myself, since I'm not authorized.

stof commented:
wrong commit message though. It is not `getItem` :smile:

stof commented:
@garak once we have phpstan running at max level in our CI, I'd rather disable scrutinizer entirely actually.

garak commented:
> wrong commit message though. It is not `getItem` smile

My bad. I hope that everything will be squashed before merging, so no big harm

garak commented:
I think I addressed all remarks.
Note that a couple of changes in `MenuManipulator` are not marked as Outdated but they actually are (I guess it's because comment is on the line before the one actually changed)

stof commented on src/Knp/Menu/MenuItem.php:
```diff
@@ -80,7 +80,7 @@ class MenuItem implements ItemInterface
      * Child items
      *
-     * @var ItemInterface[]
+     * @var array<string|int, ItemInterface>
```
keys are always strings, as they are the names

stof commented on src/Knp/Menu/MenuItem.php:
```diff
@@ -52,14 +52,14 @@ class MenuItem implements ItemInterface
      * Attributes for the item
      *
-     * @var array
+     * @var array<string, mixed>
```
for these 4 `*attributes` properties, values can be specified more precisely than `mixed`. they are `string|bool|null`

stof commented on src/Knp/Menu/MenuItem.php:
```diff
@@ -576,6 +574,10 @@ public function offsetExists($offset): bool
     /**
      * Implements ArrayAccess
+     *
+     * @param string|int $offset
```
`string`

stof replied:
same in other offsets

stof commented on src/Knp/Menu/Renderer/ArrayAccessProvider.php:
```diff
@@ -23,8 +23,9 @@ class ArrayAccessProvider implements RendererProviderInterface
     /**
-     * @param string $defaultRenderer The name of the renderer used by default
-     * @param array  $rendererIds     The map between renderer names and registry keys
+     * @param \ArrayAccess<int|string, mixed> $registry
```
keys must be strings, not integers

stof replied:
And values are `RendererInterface`

stof commented on src/Knp/Menu/Renderer/ArrayAccessProvider.php:
```diff
@@ -8,12 +8,12 @@
 {
     /**
-     * @var \ArrayAccess
+     * @var \ArrayAccess<int|string, mixed>
```
keys must be strings, not integers

stof commented on src/Knp/Menu/Renderer/TwigRenderer.php:
```diff
@@ -51,6 +51,13 @@ public function __construct(
+     *
+     * @throws \Twig\Error\LoaderError
+     * @throws \Twig\Error\RuntimeError
+     * @throws \Twig\Error\SyntaxError
```
I would remove these `@throws`

stof reviewed (changes requested):
Some of my previous feedbacks are not yet applied

garak commented:
All feedbacks applied now (and the number of errors in baseline is reduced)

stof commented on src/Knp/Menu/ItemInterface.php:
```diff
@@ -96,24 +120,44 @@ public function setChildrenAttributes(array $childrenAttributes): self;
     public function getLabelAttribute(string $name, $default = null);
 
+    /**
+     * @param mixed $value
```
@garak why ignoring this comment about the type of the value for attributes ?

garak replied:
sorry, it should ok now

garak commented:
I further reduced baseline, but I'm struggling with this one:
```
------ ----------------------------------------------------------------------------------------------------------------------------------------------- 
 Line   Knp/Menu/Util/MenuManipulator.php                                                                                                              
------ ----------------------------------------------------------------------------------------------------------------------------------------------- 
 270    Method Knp\Menu\Util\MenuManipulator::getBreadcrumbsArray() should return array<int, array('label' => string, 'uri' => string|null, 'item' =>  
        Knp\Menu\ItemInterface|null)> but returns array<int, array>.                                                                                   
------ ----------------------------------------------------------------------------------------------------------------------------------------------- 
```

stof commented:
I think it is caused by the fact that the argument is not fully described, when it can be an array (which is expected to be an array in the format of the output items)

garak commented:
So, le'ts accept it.
The last 2 errors left in baseline are that one, and another one about variable type of  `$length` argument in `slice` method of `MenuManipulator`

stof commented on src/Knp/Menu/MenuItem.php:
```diff
@@ -477,22 +485,22 @@ public function isCurrent(): ?bool
-        // if this is root, then return false
-        if ($this->isRoot()) {
+        // if this is root or there is no parent, then return false
+        if ($this->isRoot() || null === $parent = $this->getParent()) {
```
I would change that to `if (null === $this->parent)` (i.e. explicitly doing the condition done in `isRoot`) and then using `$this->parent->getLastChild()` below.

Being the root is exactly not having a parent

stof commented on src/Knp/Menu/MenuItem.php:
```diff
@@ -512,11 +520,13 @@ public function actsLikeFirst(): bool
-            // loop until we find a visible menu. If its this menu, we're first
-            if ($child->isDisplayed()) {
-                return $child->getName() === $this->getName();
+        if (null !== $parent = $this->getParent()) {
```
this can be simplified by changing the `isRoot` check above to `null === $this->parent` as well

stof commented on src/Knp/Menu/Renderer/Renderer.php:
```diff
@@ -62,8 +65,12 @@ private function htmlAttributesCallback(string $name, $value): string
      * Escapes an HTML value
      */
-    protected function escape(string $value): string
+    protected function escape(?string $value): string
```
I'd rather add a `assert($item->getUri() !== null)` in `ListRenderer::renderLinkElement` than accepting null here

stof commented on src/Knp/Menu/Util/MenuManipulator.php:
```diff
@@ -98,7 +98,7 @@ public function slice(ItemInterface $item, $offset, $length = null): ItemInterfa
 
         $slicedItem = $item->copy();
-        $children = \array_slice($slicedItem->getChildren(), $offset, $length);
+        $children = \array_slice($slicedItem->getChildren(), (int) $offset, $length);
```
I would not cast here. Instead, I would change the `is_numeric` to `is_int` above (line 86) and handle the case of `array_search` returning `false` with a proper exception saying the child does not exist

garak commented:
@stof do you think we can proceed?

stof commented on src/Knp/Menu/MenuItem.php:
```diff
@@ -24,21 +24,21 @@ class MenuItem implements ItemInterface
      * Attributes for the children list
      *
-     * @var array
+     * @var array<string, mixed>
```
link attributes and children attributes also expect `string|bool|null` for values.

And the ItemInterface should be updated to also use this type instead of `mixed`

stof commented on src/Knp/Menu/MenuItem.php:
```diff
@@ -354,6 +354,7 @@ public function reorderChildren(array $order): ItemInterface
 
         $newChildren = [];
 
+        /** @var string $name */
```
this would be better solved by documenting the argument as `string[]`

stof commented on src/Knp/Menu/MenuItem.php:
```diff
@@ -477,27 +486,27 @@ public function isCurrent(): ?bool
     {
-        // if this is root, then return false
-        if ($this->isRoot()) {
+        // if this is root , then return false
```
this space should be removed

stof commented on src/Knp/Menu/MenuItem.php:
```diff
@@ -584,6 +599,9 @@ public function offsetGet($offset)
      * Implements ArrayAccess
+     *
+     * @param string $offset
+     * @param mixed  $value
```
this is a string (or null)

stof commented on src/Knp/Menu/Provider/MenuProviderInterface.php:
```diff
@@ -9,7 +9,7 @@ interface MenuProviderInterface
      * Retrieves a menu by its name
      *
-     * @param array $options
+     * @phpstan-param array<string, callable|array{\Closure, string}> $options
```
this is wrong. Options are an `array<string, mixed>`. The type you used here was the one I asked you to use for the LazyProvider `$builders`

stof commented on src/Knp/Menu/Provider/MenuProviderInterface.php:
```diff
@@ -18,7 +18,7 @@ public function get(string $name, array $options = []): ItemInterface;
      * Checks whether a menu exists in this provider
      *
-     * @param array $options
+     * @phpstan-param array<string, callable|array{\Closure, string}> $options
```
same issue here

stof commented on src/Knp/Menu/Renderer/ListRenderer.php:
```diff
@@ -57,6 +60,10 @@ public function render(ItemInterface $item, array $options = []): string
     }
 
+    /**
+     * @param array<string, mixed> $attributes
```
attributes are `array<string, string|bool|null>`

stof commented on src/Knp/Menu/MenuItem.php:
```diff
@@ -354,7 +354,7 @@ public function reorderChildren(array $order): ItemInterface
         $newChildren = [];
 
-        /** @var string $name */
+        /** @var string[] $order */
```
put that in the phpdoc of the method instead, which documents the signature for the outside too

garak replied:
That's a problem, since method is already signed (in interface) as `array<int|string, string|self>` and here we're using array elements as string.
So, we need to change signature to `array<int|string, string>` or change some code here

stof replied:
If the implementation does not accept passing item instances, the interface should not allow it either. Otherwise, the implementation does not respect the interface.

stof commented on src/Knp/Menu/Renderer/ListRenderer.php:
```diff
@@ -45,7 +45,7 @@ public function __construct(MatcherInterface $matcher, array $defaultOptions = [
 
     /**
-     * @param array<string, mixed> $options
+     * @param array<string, string|bool|null> $options
```
please be careful when applying requested changes. I requested that for the `$attributes` argument of `renderList`, not for `$options` (where it is wrong. Options are indeed mixed)

stof commented on src/Knp/Menu/ItemInterface.php:
```diff
@@ -56,8 +57,14 @@ public function getLabel(): ?string;
     public function setLabel(?string $label): self;
 
+    /**
+     * @return array<string, mixed>
```
attributes are `string|bool|null`, not mixed.

stof commented on src/Knp/Menu/Provider/LazyProvider.php:
```diff
@@ -14,10 +14,13 @@
     private $builders;
 
+    /**
+     * @param array<string, mixed> $builders
```
@garak this comment has not been taken into account

stof commented on src/Knp/Menu/Renderer/ListRenderer.php:
```diff
@@ -194,16 +201,27 @@ protected function renderLink(ItemInterface $item, array $options = []): string
     protected function renderLinkElement(ItemInterface $item, array $options): string
     {
-        return \sprintf('<a href="%s"%s>%s</a>', $this->escape($item->getUri()), $this->renderHtmlAttributes($item->getLinkAttributes()), $this->renderLabel($item, $options));
+        $uri = null !== $item->getUri() ? $this->escape($item->getUri()) : '';
```
the uri is never null in that method (otherwise, that would render the wrong HTML anyway).

The proper fix to let phpstan know about it is to use `assert(null !== $item->getUri());` instead.

garak replied:
But we're actually allowing `$item->getUri()` to return a `null` value (and `$uri` property to be `null`)

stof replied:
in the item, yes. But when it returns `null`, we don't call `renderLinkElement` but `renderSpanElement` instead. So, in that method, it won't ever be null.

garak commented:
I've got a doubt about `$path` argument in method `get` of `Knp\Menu\Twig\Helper`: it should be casted as `array<int, string>`? Because it's then used in a foreach loop as such. But then method `castMenu` in same class is calling it with `array<Knp\Menu\ItemInterface|string>` instead

stof commented:
Actually, in `castMenu`, only the first element of the array can be `ItemInterface|string`. All others are expected to be strings.

garak commented on src/Knp/Menu/Twig/MenuExtension.php:
```diff
@@ -12,8 +12,19 @@
     private $helper;
+
+    /**
+     * @var MatcherInterface|null
```
it is, but I experienced that this can be useful in the future if we want programmatically add typed properties (e.g. using rector)

garak commented:
Hey, is this good to merge now? It looks like everything was addressed

stof commented on src/Knp/Menu/Twig/Helper.php:
```diff
@@ -104,7 +119,7 @@ public function render($menu, array $options = [], ?string $renderer = null): st
      * @param mixed $subItem A string or array to append onto the end of the array
      *
-     * @return array
+     * @return array<int, array<string, mixed>>
```
this could be more precised, by reusing the type of the MenuManipulator method

stof commented on src/Knp/Menu/Twig/MenuExtension.php:
```diff
@@ -72,10 +94,10 @@ public function render($menu, array $options = [], ?string $renderer = null): st
+     * @param string|array<string|null>|null                   $subItem
      *
-     * @return array
+     * @return array<int, array<string, mixed>>
```
same here

stof commented on src/Knp/Menu/Util/MenuManipulator.php:
```diff
@@ -197,9 +206,10 @@ public function toArray(ItemInterface $item, ?int $depth = null): array
      *   * [['label' => 'subItem1', 'url' => '@homepage'], ['label' => 'subItem2']]
      *
-     * @param mixed $subItem A string or array to append onto the end of the array
+     * @param string|ItemInterface|array<int|string, mixed>|\Traversable<mixed> $subItem A string or array to append onto the end of the array
```
We could add a more precise type for the array and iterator instead of `mixed`:

```
@phpstan-param string|ItemInterface|array<int|string, string|int|float|null|array{label: string, url: string|null, item: ItemInterface|null}|ItemInterface>|\Traversable<string|int|float|null|array{label: string, url: string|null, item: ItemInterface|null}|ItemInterface> $subItem
```

stof replied:
and then, this should also be applied to callers (the templating helper and the twig extension).
This should remove the need for the baseline entry on that method.

garak replied:
Well, it looks like is not enough. PHPStan is still complaining

stof commented on src/Knp/Menu/Util/MenuManipulator.php:
```diff
@@ -276,6 +290,10 @@ private function buildBreadcrumbsArray(ItemInterface $item): array
     }
 
+    /**
+     * @return array<string, string|ItemInterface|null>
```
I would remove that one, keeping only the `@phpstan-return` and the `array` return type. `array<string, string|ItemInterface|null>` would create confusion as it would imply that any of the keys could be of the type `string|ItemInterface|null`.
PHPStan won't complain about an unspecified array type, as it is specified in `@phpstan-return`.

stof commented on src/Knp/Menu/Util/MenuManipulator.php:
```diff
@@ -265,6 +275,10 @@ public function getBreadcrumbsArray(ItemInterface $item, $subItem = null): array
     }
 
+    /**
+     * @return array<int, array<string, string|ItemInterface|null>>
```
I would remove that one, keeping only the `@phpstan-return` and the `array` return type. `array<string, string|ItemInterface|null>` would create confusion as it would imply that any of the keys could be of the type `string|ItemInterface|null`.
PHPStan won't complain about an unspecified array type, as it is specified in `@phpstan-return`.
